### ISSUE TITLE
<complex.h> and <complex> interoperability from C++17 onwards

### ISSUE THREAD

**Issue by FabienPean** [microsoft/STL#3280]
**TLDR;** consuming a C library relying on recommended typedefs in `complex.h` from C++ compiling with standard 17 onwards breaks build.

From C++17, the C header `<complex.h>` is basically replaced by the C++ `<complex>` according to the [standard](https://eel.is/c++draft/complex.h.syn)

UCRT and STL defines a common structure for complex numbers as such:

https://github.com/microsoft/STL/blob/4483e87f6e5aa5c4598404fc01eecb7e547dd439/stl/inc/complex#L42-L56

However, the [documentation of the UCRT for complex numbers](https://learn.microsoft.com/en-us/cpp/c-runtime-library/complex-math-support) does not mention the structure above and recommends using the following typedefs: `_Fcomplex`, `_Dcomplex`, and `_Lcomplex`

When using a standard below C++17, `<complex.h>` is used which defines those typedefs but nonexistent in the STL implementation. So consuming C headers relying on recommended types from UCRT from a C++ compiler breaks build when simply switching flags for standard. 

Note that the STL implementation declares some closely related typedefs are wrapped in the std namespace https://github.com/microsoft/STL/blob/4483e87f6e5aa5c4598404fc01eecb7e547dd439/stl/inc/complex#L347-L349

The use of `<complex.h>` can be forced by adding a compile definition `_CRT_USE_C_COMPLEX_H` that can be found browsing the header, for which I could not find information online.

Is the nature of the problem actually STL related or to be solved elsewhere (documentation, error message,etc)?

**Comment by strega-nil-ms** [comment]
I'm not sure on the exact answer; most of the people on this team are on vacation right now, so hopefully we can get back to you after the new year 😸

**Comment by frederick-vs-ja** [comment]
This seems an open issue in C & C++ standards (although MSVC hasn't implemented C's `_Complex` yet). While the binary compatibility between C's `_Complex` and C++'s `std::complex` is intended, the current standards haven't offered source compatibility.

**Comment by h-vetinari** [comment]
Could this issue be reopened? There's no interoperability, and the private workaround is still undocumented. Also quoting my comment from the PR:
> > We don't support C99's `_Complex`
> 
> Are there plans (on any finite timescale) to change that? I know this repo is for the C++ stdlib, not the C one, but as someone helping maintain scipy and 100+ packages in conda-forge, dealing with C-complex code on windows is a constant source of pain.
> 
> This issue now also breaks lapack-detection in opencv when using C++17, which is the sort of action-at-a-distance that nobody is prepared for, and correspondingly takes a lot of digging to unearth.
> 
> Which is all extra-exasperating because there _is_ [support](https://learn.microsoft.com/en-us/cpp/c-runtime-library/complex-math-support) for much of the C99 interface, just without the standardized types.
> 
> There was a [tweet](https://twitter.com/VisualStudio/status/1612864351186161692) at the beginning of the year that promised complex support, but all the release notes since then have been silent on this.
> 
> PS. I'm aware of the developercommunity [issue](https://developercommunity.visualstudio.com/t/714008) for this...
